src/fu-engine-repair.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * Copyright (C) 2023 Kate Hsuan <hpa@redhat.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuEngine"
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <glib/gi18n.h>
+#include <glib/gstdio.h>
+#ifdef HAVE_GIO_UNIX
+#include <gio/gunixfdlist.h>
+#include <glib-unix.h>
+#endif
+#include <locale.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "fu-console.h"
+#include "fu-engine-repair.h"
+#include "fu-engine.h"
+#include "fu-plugin-private.h"
+#include "fu-security-attr-common.h"
+#include "fu-security-attrs-private.h"
+#include "fu-util-common.h"
+
+static gboolean
+is_grubby_installed(GError **error)
+{
+	g_autofree gchar *grubby = NULL;
+
+	grubby = fu_path_find_program("grubby", error);
+	if (!grubby)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+grubby_set(gboolean enable, const gchar *grubby_arg, GError **error)
+{
+	g_autofree gchar *output = NULL;
+	g_autofree gchar *arg_string = NULL;
+	const gchar *argv_grubby[] = {"", "--update-kernel=DEFAULT", "", NULL};
+	g_autofree gchar *grubby = NULL;
+
+	grubby = fu_path_find_program("grubby", NULL);
+	argv_grubby[0] = grubby;
+
+	if (enable)
+		arg_string = g_strdup_printf("--args=%s", grubby_arg);
+	else
+		arg_string = g_strdup_printf("--remove-args=%s", grubby_arg);
+
+	argv_grubby[2] = arg_string;
+
+	if (!g_spawn_sync(NULL,
+			  (gchar **)argv_grubby,
+			  NULL,
+			  G_SPAWN_DEFAULT,
+			  NULL,
+			  NULL,
+			  &output,
+			  NULL,
+			  NULL,
+			  error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+is_kernel_parameter(const gchar *para_string, GError **error)
+{
+	g_autofree gchar *kernel_cmdline = NULL;
+	gsize length;
+
+	if (!g_file_get_contents("/proc/cmdline", &kernel_cmdline, &length, error)) {
```

Comment:
Don't we have a helper for kernel command line already?

Reply:
Thank you for pointing it out. I found it (fu_kernel_parse_cmdline) :).

Reply:
By the way, should I move IOMMU kernel parameter check to fu-kernel.c? Since the kernel lockdown test was already placed there.